File: examples/python/s3_example.py
#!/usr/bin/env python3
"""
Example script demonstrating how to use AWS S3 Simulator with boto3.
Install required package: pip install boto3
"""

import os
import boto3
from botocore.exceptions import ClientError


def create_s3_client():
    """Create and return an S3 client configured for local MinIO."""
    return boto3.client(
        's3',
        endpoint_url=os.getenv('S3_ENDPOINT', 'http://localhost:9000'),
        aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID', 'admin'),
        aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY', 'password123'),
        region_name=os.getenv('AWS_DEFAULT_REGION', 'us-east-1'),
    )


def list_buckets(s3_client):
    """List all S3 buckets."""
    try:
        response = s3_client.list_buckets()
        print("\n📦 Available buckets:")
        for bucket in response['Buckets']:
            print(f"  - {bucket['Name']}")
        return response['Buckets']
    except ClientError as e:
        print(f"Error listing buckets: {e}")
        return []


def create_bucket(s3_client, bucket_name):
    """Create a new S3 bucket."""
    try:
        s3_client.create_bucket(Bucket=bucket_name)
        print(f"✓ Bucket '{bucket_name}' created successfully")
    except ClientError as e:
        if e.response['Error']['Code'] == 'BucketAlreadyOwnedByYou':
            print(f"ℹ Bucket '{bucket_name}' already exists")
        else:
            print(f"Error creating bucket: {e}")


def upload_file(s3_client, bucket_name, file_path, object_name=None):
    """Upload a file to S3 bucket."""
    if object_name is None:
        object_name = os.path.basename(file_path)
    
    try:
        s3_client.upload_file(file_path, bucket_name, object_name)
        print(f"✓ File '{file_path}' uploaded as '{object_name}'")
    except ClientError as e:
        print(f"Error uploading file: {e}")


def list_objects(s3_client, bucket_name):
    """List all objects in a bucket."""
    try:
        response = s3_client.list_objects_v2(Bucket=bucket_name)
        
        if 'Contents' not in response:
            print(f"📭 Bucket '{bucket_name}' is empty")
            return []
        
        print(f"\n📄 Objects in '{bucket_name}':")
        for obj in response['Contents']:
            size_kb = obj['Size'] / 1024
            print(f"  - {obj['Key']} ({size_kb:.2f} KB)")
        
        return response['Contents']
    except ClientError as e:
        print(f"Error listing objects: {e}")
        return []


def download_file(s3_client, bucket_name, object_name, local_path):
    """Download a file from S3 bucket."""
    try:
        s3_client.download_file(bucket_name, object_name, local_path)
        print(f"✓ File '{object_name}' downloaded to '{local_path}'")
    except ClientError as e:
        print(f"Error downloading file: {e}")


def delete_object(s3_client, bucket_name, object_name):
    """Delete an object from S3 bucket."""
    try:
        s3_client.delete_object(Bucket=bucket_name, Key=object_name)
        print(f"✓ Object '{object_name}' deleted")
    except ClientError as e:
        print(f"Error deleting object: {e}")


def main():
    """Main function demonstrating S3 operations."""
    print("=" * 60)
    print("AWS S3 Simulator - Python Example (boto3)")
    print("=" * 60)
    
    # Create S3 client
    s3 = create_s3_client()
    
    # List existing buckets
    list_buckets(s3)
    
    # Create a test bucket
    test_bucket = 'python-test-bucket'
    create_bucket(s3, test_bucket)
    
    # Create a sample file
    sample_file = 'test_file.txt'
    with open(sample_file, 'w') as f:
        f.write('Hello from AWS S3 Simulator!\n')
        f.write('This is a test file created by Python example.\n')
    
    # Upload file
    upload_file(s3, test_bucket, sample_file)
    
    # List objects in bucket
    list_objects(s3, test_bucket)
    
    # Download file
    download_path = 'downloaded_file.txt'
    download_file(s3, test_bucket, sample_file, download_path)
    
    # Clean up local files
    if os.path.exists(sample_file):
        os.remove(sample_file)
    if os.path.exists(download_path):
        os.remove(download_path)
    
    print("\n" + "=" * 60)
    print("✓ Example completed successfully!")
    print("=" * 60)


if __name__ == '__main__':
    main()

File: examples/nodejs/s3_example.js
#!/usr/bin/env node
/**
 * Example script demonstrating how to use AWS S3 Simulator with AWS SDK v3.
 * Install required packages:
 * npm install @aws-sdk/client-s3
 */

const fs = require('fs');
const {
  S3Client,
  ListBucketsCommand,
  CreateBucketCommand,
  PutObjectCommand,
  ListObjectsV2Command,
  GetObjectCommand,
  DeleteObjectCommand,
} = require('@aws-sdk/client-s3');

// Configure S3 client for local MinIO
const s3Client = new S3Client({
  endpoint: process.env.S3_ENDPOINT || 'http://localhost:9000',
  region: process.env.AWS_DEFAULT_REGION || 'us-east-1',
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID || 'admin',
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || 'password123',
  },
  forcePathStyle: true, // Required for MinIO
});

/**
 * List all S3 buckets
 */
async function listBuckets() {
  try {
    const response = await s3Client.send(new ListBucketsCommand({}));
    console.log('\n📦 Available buckets:');
    response.Buckets.forEach((bucket) => {
      console.log(`  - ${bucket.Name}`);
    });
    return response.Buckets;
  } catch (error) {
    console.error('Error listing buckets:', error.message);
    return [];
  }
}

/**
 * Create a new S3 bucket
 */
async function createBucket(bucketName) {
  try {
    await s3Client.send(new CreateBucketCommand({ Bucket: bucketName }));
    console.log(`✓ Bucket '${bucketName}' created successfully`);
  } catch (error) {
    if (error.name === 'BucketAlreadyOwnedByYou') {
      console.log(`ℹ Bucket '${bucketName}' already exists`);
    } else {
      console.error('Error creating bucket:', error.message);
    }
  }
}

/**
 * Upload a file to S3 bucket
 */
async function uploadFile(bucketName, filePath, objectKey) {
  try {
    const fileContent = fs.readFileSync(filePath);
    await s3Client.send(
      new PutObjectCommand({
        Bucket: bucketName,
        Key: objectKey,
        Body: fileContent,
      })
    );
    console.log(`✓ File '${filePath}' uploaded as '${objectKey}'`);
  } catch (error) {
    console.error('Error uploading file:', error.message);
  }
}

/**
 * List all objects in a bucket
 */
async function listObjects(bucketName) {
  try {
    const response = await s3Client.send(
      new ListObjectsV2Command({ Bucket: bucketName })
    );

    if (!response.Contents || response.Contents.length === 0) {
      console.log(`📭 Bucket '${bucketName}' is empty`);
      return [];
    }

    console.log(`\n📄 Objects in '${bucketName}':`);
    response.Contents.forEach((obj) => {
      const sizeKB = (obj.Size / 1024).toFixed(2);
      console.log(`  - ${obj.Key} (${sizeKB} KB)`);
    });

    return response.Contents;
  } catch (error) {
    console.error('Error listing objects:', error.message);
    return [];
  }
}

/**
 * Download a file from S3 bucket
 */
async function downloadFile(bucketName, objectKey, localPath) {
  try {
    const response = await s3Client.send(
      new GetObjectCommand({
        Bucket: bucketName,
        Key: objectKey,
      })
    );

    const stream = response.Body;
    const writeStream = fs.createWriteStream(localPath);

    stream.pipe(writeStream);

    await new Promise((resolve, reject) => {
      writeStream.on('finish', resolve);
      writeStream.on('error', reject);
    });

    console.log(`✓ File '${objectKey}' downloaded to '${localPath}'`);
  } catch (error) {
    console.error('Error downloading file:', error.message);
  }
}

/**
 * Delete an object from S3 bucket
 */
async function deleteObject(bucketName, objectKey) {
  try {
    await s3Client.send(
      new DeleteObjectCommand({
        Bucket: bucketName,
        Key: objectKey,
      })
    );
    console.log(`✓ Object '${objectKey}' deleted`);
  } catch (error) {
    console.error('Error deleting object:', error.message);
  }
}

/**
 * Main function demonstrating S3 operations
 */
async function main() {
  console.log('='.repeat(60));
  console.log('AWS S3 Simulator - Node.js Example (AWS SDK v3)');
  console.log('='.repeat(60));

  try {
    // List existing buckets
    await listBuckets();

    // Create a test bucket
    const testBucket = 'nodejs-test-bucket';
    await createBucket(testBucket);

    // Create a sample file
    const sampleFile = 'test_file.txt';
    const content = [
      'Hello from AWS S3 Simulator!',
      'This is a test file created by Node.js example.',
    ].join('\n');
    fs.writeFileSync(sampleFile, content);

    // Upload file
    await uploadFile(testBucket, sampleFile, sampleFile);

    // List objects in bucket
    await listObjects(testBucket);

    // Download file
    const downloadPath = 'downloaded_file.txt';
    await downloadFile(testBucket, sampleFile, downloadPath);

    // Clean up local files
    if (fs.existsSync(sampleFile)) {
      fs.unlinkSync(sampleFile);
    }
    if (fs.existsSync(downloadPath)) {
      fs.unlinkSync(downloadPath);
    }

    console.log('\n' + '='.repeat(60));
    console.log('✓ Example completed successfully!');
    console.log('='.repeat(60));
  } catch (error) {
    console.error('Error in main:', error);
    process.exit(1);
  }
}

// Run the example
main();